modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/adapter/SW360ReleaseClientAdapter.java

```diff
@@ -12,71 +12,121 @@
 import org.eclipse.sw360.antenna.sw360.rest.resource.SW360HalResourceUtility;
 import org.eclipse.sw360.antenna.sw360.rest.resource.components.SW360Component;
+import org.eclipse.sw360.antenna.sw360.rest.resource.components.SW360ComponentEmbedded;
 import org.eclipse.sw360.antenna.sw360.rest.resource.releases.SW360Release;
 import org.eclipse.sw360.antenna.sw360.rest.resource.releases.SW360SparseRelease;
+import org.eclipse.sw360.antenna.sw360.utils.SW360ComponentAdapterUtils;
 import org.eclipse.sw360.antenna.sw360.utils.SW360ReleaseAdapterUtils;
 import org.eclipse.sw360.antenna.util.ProxySettings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.http.HttpHeaders;
 
 import java.nio.file.Path;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
-import java.util.Set;
 
 public class SW360ReleaseClientAdapter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SW360ReleaseClientAdapter.class);
+
     private final SW360ReleaseClient releaseClient;
+    private final SW360ComponentClientAdapter sw360ComponentClientAdapter;
 
     public SW360ReleaseClientAdapter(String restUrl, ProxySettings proxySettings) {
         this.releaseClient = new SW360ReleaseClient(restUrl, proxySettings);
+        sw360ComponentClientAdapter = new SW360ComponentClientAdapter(restUrl, proxySettings);
     }
 
-    public SW360Release addRelease(Artifact artifact, SW360Component sw360Component, Set<String> sw360LicenseIds, boolean uploadSource, HttpHeaders header) {
-        SW360Release release = SW360ReleaseAdapterUtils.prepareRelease(sw360Component, sw360LicenseIds, artifact);
+    public SW360Release getOrCreateRelease(SW360Release sw360ReleaseFromArtifact, HttpHeaders header, boolean uploadSources, boolean updateReleases) {
+        return getRelease(sw360ReleaseFromArtifact, header)
+                .map(sw360ReleaseFromArtifact::mergeWith) // TODO: always merge?
+                .map(sw360Release -> {
+                    if(updateReleases) {
+                        return releaseClient.patchRelease(sw360Release, header);
+                    } else {
+                        return sw360Release;
+                    }
+                })
+                .orElseGet(() -> createRelease(sw360ReleaseFromArtifact, uploadSources, header));
+    }
 
-        if (! SW360ReleaseAdapterUtils.isValidRelease(release)) {
-            throw new ExecutionException("Can not write invalid release for " + artifact.toString());
+    /*
+     * Create a release in SW360
+     */
+    public SW360Release createRelease(SW360Release releaseFromArtifact, boolean uploadSource, HttpHeaders header) {
+        if (! SW360ReleaseAdapterUtils.isValidRelease(releaseFromArtifact)) {
+            throw new ExecutionException("Can not write invalid release for " + releaseFromArtifact.getName() + "-" + releaseFromArtifact.getVersion());
+        }
+        if (releaseFromArtifact.getReleaseId() != null) {
```

Comment:
doesn't this completely eliminate the updating function?

Reply:
This is within `createRelease`, which just creates nonexisting releases. updating is handled somewhere else.